src/Tiny_httpd.ml

```diff
@@ -83,6 +83,27 @@ module Byte_stream = struct
   let of_chan = of_chan_ ~close:close_in
   let of_chan_close_noerr = of_chan_ ~close:close_in_noerr
 
+  exception Timeout
+
+  let of_descr_ ?(timeout=(-1.0)) ~close ic : t =
+    let i = ref 0 in
+    let len = ref 0 in
+    let buf = Bytes.make 4096 ' ' in
+    { bs_fill_buf=(fun () ->
+      if !i >= !len then (
+        i := 0;
+        let (to_read,_,_) = Unix.select [ic] [] [] timeout in
+        if to_read = [] then raise Timeout;
+        try len := Unix.read ic buf 0 (Bytes.length buf)
```

Comment:
btw, should this be in a loop?

Reply:
Our file descriptor are sockets with an internal default buffer size of:
On 64-bit Linux 4.14.18: 212992 bytes
On 32-bit Linux 4.4.92: 163840 bytes

So if read returns less than those value, it means that the writer has currently written less that this number of bytes and the best is to return so that the scanning of the request can proceed, while the writer may continue to write, if the request was not send completely yet. 

For a real file, read will "almost always" return the number of bytes required (except at the end of file). 

The best would be to use the same buffer size for our buffer and the socket buffer using (or ensure at least a greater size)

val getsockopt_int : file_descr -> socket_int_option -> int
(** Same as {!getsockopt} for an integer-valued socket option. *)

val setsockopt_int : file_descr -> socket_int_option -> int -> unit
(** Same as {!setsockopt} for an integer-valued socket option. *)

Reply:
sorry, I mean the `try`. Should it be in a loop so that in case of exception (`EAGAIN`) we try to read again?